README
ZeroMQ-Oneliner

This module can be used to easily create ZeroMQ sockets without need to
call zmq_setsockopt, bind, connect functions. All socket options is described
in same way as http url does.

Example socket:
	
	bind-tcp://*:6666/pull?hwm=100

This means that we need to bind tcp socket on port 6666, which is ZMQ_PULL
and have high water mark set to 100.

Possible variations:
	- connection method:
		- bind
		- connect
	- connection type:
		- tcp
		- inproc
		- ipc
		- pgm
		- epgm
	- address - same as zeromq supports
	- zmq socket type:
		- push
		- pull
		- rep
		- req
		- pub
		- sub
	- options:
		- hwm
		- swap
		- identity
		- subscribe
		- rate
		- recovery
		- sndbuf
		- rcvbuf
		- linger
		- reconnect
		- backlog

CONSOLE TOOLS

There are two console tools included:
	- zmq_send wait for input on stdin and send each incoming line as separate message.
	- zmq_recv print all incoming messages into stdout.

Both called same way:
	
	zmq_send connect-tcp://127.0.0.1:6666/push

Currently only pub/sub and push/pull supported for console tools. Module support all
types.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc ZeroMQ::Oneliner

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=ZeroMQ-Oneliner

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/ZeroMQ-Oneliner

    CPAN Ratings
        http://cpanratings.perl.org/d/ZeroMQ-Oneliner

    Search CPAN
        http://search.cpan.org/dist/ZeroMQ-Oneliner/


LICENSE AND COPYRIGHT

Copyright (C) 2012 x86

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this program.  If not, see
L<http://www.gnu.org/licenses/>.



Changes
Revision history for ZeroMQ-Oneliner

0.01    Date/time
        First version, released on an unsuspecting world.



Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'ZeroMQ::Oneliner',
    AUTHOR           => q{x86 <x86mail@gmail.com>},
    VERSION_FROM     => 'lib/ZeroMQ/Oneliner.pm',
    ABSTRACT_FROM    => 'lib/ZeroMQ/Oneliner.pm',
    LICENSE          => 'LGPL_3',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        #'ABC'              => 1.6,
        #'Foo::Bar::Module' => 5.0401,
	'ZeroMQ' => 0.23,
	'Getopt::Lazy' => '0.0.6',
    },
    EXE_FILES => [ 'bin/zmq_send', 'bin/zmq_recv' ],
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'ZeroMQ-Oneliner-*' },
);


lib/ZeroMQ/Oneliner.pm
package ZeroMQ::Oneliner;

use 5.006;
use strict;
use warnings FATAL => 'all';
use ZeroMQ qw(:all);

=head1 NAME

ZeroMQ::Oneliner - The great new ZeroMQ::Oneliner!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

our $types = {
	"push" => ZMQ_PUSH,
	"pull" => ZMQ_PULL,
	"req"  => ZMQ_REQ,
	"rep"  => ZMQ_REP,
	"pub"  => ZMQ_PUB,
	"sub"  => ZMQ_SUB,
};
our $opts = {
	"hwm"         => ZMQ_HWM,
	"swap"        => ZMQ_SWAP,
	"identity"    => ZMQ_IDENTITY,
	"subscribe"   => ZMQ_SUBSCRIBE,
	"rate"        => ZMQ_RATE,
	"recovery"    => ZMQ_RECOVERY_IVL,
	"sndbuf"      => ZMQ_SNDBUF,
	"rcvbuf"      => ZMQ_RCVBUF,
	"linger"      => ZMQ_LINGER,
	"reconnect"   => ZMQ_RECONNECT_IVL,
	"backlog"     => ZMQ_BACKLOG,
};

my $ctx = ZeroMQ::Context->new();

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use ZeroMQ::Oneliner;

    my $foo = ZeroMQ::Oneliner->new("bind-tcp://*:6667/push?hwm=100");
    print $foo "send message";
    
    my $bar = ZeroMQ::Oneliner->new("connect-tcp://*:6667/pull");
    print <$bar>; # recv message

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub TIEHANDLE {
    (
        ( defined( $_[1] ) && UNIVERSAL::isa( $_[1], __PACKAGE__ ) )
        ? $_[1]
        : shift->new(@_)
    );	
}

sub PRINT {    my $self = shift; *$self->{socket}->send(@_);      } 
sub READLINE { my $self = shift; my $msg = *$self->{socket}->recv(); $msg->data(); }
sub CLOSE {    my $self = shift; *$self->{socket}->close();      }

sub new {
	my ($class, $name) = @_;
	
	my $self = bless \do { local *FH }, $class;
	tie *$self, $class, $self;
	
	my ($socket);
	
	if($name =~ m@(bind|connect)-((?:inproc|ipc|tcp|pgm|epgm)://[[:alnum:].*]+:[0-9]+)/(re[qp]|push|pull|[ps]ub)[?]?([[:alnum:]=&]+)?@i){
		my ($bc, $zmq_address, $s_type, $s_options) = ($1, $2, $3, $4);
		
		$socket = $ctx->socket($types->{$s_type});
		
		if($s_options){
			foreach my $option (split /&/, $s_options){
				my ($k, $v) = split /=/, $option;
				
				$k = $opts->{$k};
				$v = int($v) if $v =~ /^\d+$/;
				
				$socket->setsockopt($k, $v);
			}
		}
		
		if($bc eq "bind"){    $socket->bind($zmq_address);    }
		if($bc eq "connect"){ $socket->connect($zmq_address); }
		
		*$self->{socket} = $socket;
	}
	return $self;
}

=head1 AUTHOR

x86, C<< <x86mail at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-zeromq-oneliner at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=ZeroMQ-Oneliner>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc ZeroMQ::Oneliner


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=ZeroMQ-Oneliner>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/ZeroMQ-Oneliner>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/ZeroMQ-Oneliner>

=item * Search CPAN

L<http://search.cpan.org/dist/ZeroMQ-Oneliner/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012 x86.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this program.  If not, see
L<http://www.gnu.org/licenses/>.


=cut

1; # End of ZeroMQ::Oneliner
